internal/pkg/manifests/query/builder.go

```diff
@@ -0,0 +1,252 @@
+package query
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/thanos-community/thanos-operator/internal/pkg/manifests"
+)
+
+const (
+	// Name is the name of the Thanos Query component.
+	Name = "thanos-query"
+
+	// ComponentName is the name of the Thanos Query component.
+	ComponentName = "query-layer"
+
+	GRPCPort     = 10901
+	GRPCPortName = "grpc"
+
+	HTTPPort     = 9090
+	HTTPPortName = "http"
+)
+
+// QuerierOptions for Thanos Querier
+type QuerierOptions struct {
+	manifests.Options
+	ReplicaLabels []string
+	Timeout       string
+	LookbackDelta string
+	MaxConcurrent int
+
+	Endpoints []Endpoint
+}
+
+type EndpointType string
+
+const (
+	Regular     EndpointType = "regular"
```

Comment:
maybe we could have an ApplyDefault func here and set one of these?

Reply:
Yup this will be mostly based on watch logic, but plan to do something around that.